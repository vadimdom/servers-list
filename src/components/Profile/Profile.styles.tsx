import { styled } from 'styled-components';

export const ProfileContainer = styled.div`
  display: flex;
  align-items: center;
  gap: 10px;
  font-weight: bold;
`;

export const ImageAndNameContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: 12px;
`;

export const ProfileImage = styled.img`
  width: 30px;
  height: 30px;
  border-radius: 50%;
  border: 1px solid ${({ theme }) => theme.color.blue};
`;

export const LogoutContainer = styled.span`
  cursor: pointer;
  transition: color 0.2s linear;

  &:hover {
    color: ${({ theme }) => theme.color.blue};
  }
`;
