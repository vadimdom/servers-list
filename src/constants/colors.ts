export const COLORS = {
  blue: '#1596eb',
  lightBlue: '#5bb3ee',
  gray: '#cfdbe6',
  white: '#ffffff',
};
