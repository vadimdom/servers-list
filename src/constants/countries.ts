import { Country } from '../types';

export const AVAILABLE_COUNTRY_CODES = ['UK', 'US', 'DE', 'JA'] as const;

export const COUNTRY_FULL_NAMES: { [key in Country]: string } = {
  UK: 'united kingdom',
  US: 'united states',
  DE: 'germany',
  JA: 'japan',
};
