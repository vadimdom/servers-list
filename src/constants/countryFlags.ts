import deFlag from './img/de.svg';
import jaFlag from './img/ja.svg';
import usFlag from './img/us.svg';
import ukFlag from './img/uk.svg';

import { Country } from '../types';

export const COUNTRY_FLAGS: { [key in Country]: string } = {
  UK: ukFlag,
  US: usFlag,
  DE: deFlag,
  JA: jaFlag,
};
