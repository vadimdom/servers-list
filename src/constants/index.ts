export { DEVICE_SIZES } from './deviceSizes';
export { COLORS } from './colors';
export { FALLBACK_IMAGE, DEFAULT_PROFILE_IMAGE, BROKEN_DEFAULT_PROFILE_IMAGE } from './profileImage';
export { MIN_USERNAME_LENGTH, MIN_PASSWORD_LENGTH } from './validation';
export { AVAILABLE_COUNTRY_CODES, COUNTRY_FULL_NAMES } from './countries';
export { COUNTRY_FLAGS } from './countryFlags';
