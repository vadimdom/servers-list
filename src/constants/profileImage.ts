export const FALLBACK_IMAGE =
  'https://cdn.vectorstock.com/i/preview-1x/65/30/default-image-icon-missing-picture-page-vector-40546530.jpg';

export const DEFAULT_PROFILE_IMAGE =
  'https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png';

export const BROKEN_DEFAULT_PROFILE_IMAGE = 'https://cdn.pixabayk-profile-picture-973460_1280.png';
