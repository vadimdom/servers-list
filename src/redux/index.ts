export { selectUser, setUser, removeUser } from './profileSlice';
export { useAppDispatch, useAppSelector } from './hooks';
export { ReduxStoreProvider } from './provider';
