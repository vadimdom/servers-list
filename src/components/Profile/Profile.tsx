import { useNavigate } from 'react-router-dom';
import { SyntheticEvent } from 'react';

import { useAppDispatch, useAppSelector, removeUser, selectUser } from '../../redux';
import { DEFAULT_PROFILE_IMAGE, FALLBACK_IMAGE } from '../../constants';
import { ProfileImage, LogoutContainer, ProfileContainer, ImageAndNameContainer } from './Profile.styles';

export const Profile = () => {
  const navigate = useNavigate();
  const dispatch = useAppDispatch();
  const user = useAppSelector(selectUser);

  const onLogout = () => {
    localStorage.removeItem('token');
    dispatch(removeUser());
    navigate('/');
  };

  return (
    <ProfileContainer>
      <LogoutContainer data-cy="logout" onClick={onLogout}>
        Logout
      </LogoutContainer>
      <ImageAndNameContainer>
        <ProfileImage
          data-testid="profile-image"
          src={user?.image || DEFAULT_PROFILE_IMAGE}
          onError={(e: SyntheticEvent<HTMLImageElement, Event>) => (e.currentTarget.src = FALLBACK_IMAGE)}
        />
        <span>{user?.name}</span>
      </ImageAndNameContainer>
    </ProfileContainer>
  );
};
