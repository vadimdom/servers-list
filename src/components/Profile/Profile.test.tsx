import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import 'jest-styled-components';

import { Profile } from './Profile';
import { ThemeProvider } from '../../theme';
import { BROKEN_DEFAULT_PROFILE_IMAGE, FALLBACK_IMAGE, DEFAULT_PROFILE_IMAGE } from '../../constants';
import { useAppSelector } from '../../redux';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

const mockedUseDispatch = jest.fn();

jest.mock('../../redux', () => ({
  ...jest.requireActual('../../redux'),
  useAppDispatch: () => mockedUseDispatch,
  useAppSelector: jest.fn(),
}));

const renderProfile = () => {
  render(
    <MemoryRouter>
      <ThemeProvider>
        <Profile />
      </ThemeProvider>
    </MemoryRouter>
  );
};

const localStorageMock = (() => {
  return {
    removeItem: () => null,
  };
})();

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock,
});

describe('<Profile />', () => {
  beforeEach(async () => {
    (useAppSelector as jest.Mock).mockReturnValue({ name: 'User', image: DEFAULT_PROFILE_IMAGE });
  });
  afterEach(() => {
    (useAppSelector as jest.Mock).mockClear();
  });

  test('should logout on logout click', async () => {
    const localStorageRemoveItem = jest.spyOn(localStorage, 'removeItem');
    renderProfile();

    expect(screen.getByTestId('profile')).toMatchSnapshot();
    fireEvent.click(screen.getByText('Logout'));

    await waitFor(() => expect(localStorageRemoveItem).toBeCalledWith('token'));
    expect(mockedUseDispatch).toBeCalledWith({ payload: undefined, type: 'profile/removeUser' });
    expect(mockedUseNavigate).toBeCalledWith('/');
  });
  test('should display correct user image', () => {
    renderProfile();

    expect(useAppSelector).toHaveBeenCalled();
    expect((screen.getByTestId('profile-image') as HTMLImageElement).src).toBe(DEFAULT_PROFILE_IMAGE);
  });
  test('should display fallback image, if user image is broken', () => {
    (useAppSelector as jest.Mock).mockReturnValue({ name: 'User', image: BROKEN_DEFAULT_PROFILE_IMAGE });

    renderProfile();

    fireEvent.error(screen.getByTestId('profile-image'));

    expect((screen.getByTestId('profile-image') as HTMLImageElement).src).toBe(FALLBACK_IMAGE);
  });
});
