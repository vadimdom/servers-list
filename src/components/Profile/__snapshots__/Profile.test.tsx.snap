// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile /> should logout on logout click 1`] = `
.c0 {
  display: flex;
  align-items: center;
  gap: 10px;
  font-weight: bold;
}

.c2 {
  display: flex;
  flex-direction: column;
  align-items: center;
  font-size: 12px;
}

.c3 {
  width: 30px;
  height: 30px;
  border-radius: 50%;
  border: 1px solid #1596eb;
}

.c1 {
  cursor: pointer;
  transition: color 0.2s linear;
}

.c1:hover {
  color: #1596eb;
}

<div
  class="c0"
  data-testid="profile"
>
  <span
    class="c1"
    data-cy="logout"
  >
    Logout
  </span>
  <div
    class="c2"
  >
    <img
      class="c3"
      data-testid="profile-image"
      src="https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_1280.png"
    />
    <span>
      User
    </span>
  </div>
</div>
`;
